### DISCUSSION TITLE
Message latency depends on the workload of the process owning the data writer

### DISCUSSION THREAD

**Discussion by MMarcus95** [eProsima/Fast-DDS#5542] [General]
### Is there an already existing issue for this?

- [X] I have searched the existing issues

### Expected behavior

The time between message publication and message reception does not depend on the workload of the of the process owning the data writer

### Current behavior

The time between message publication and message reception depends on the workload of the of the process owning the data writer

### Steps to reproduce

I'm testing the message latency between a data writer and a data reader. They belong to separate processes, and I'm using the  discovery server, launching the server in a third process. To publish the message I'm using the following while loop

```
RoundTripTimeMsg msg;
while (true)
{
    msg.id() += 1;
    msg.timestamp() = std::chrono::duration_cast<std::chrono::nanoseconds>(std::chrono::high_resolution_clock::now().time_since_epoch()).count();
    writer->write(&msg);
    
    for (int i = 0; i < 10000000; i++) {
        // Simulating workload
    }
}
```
The for loop simulates a workload. RoundTripTimeMsg is instead a custom message defined as follows

```
struct RoundTripTimeMsg{
    unsigned long long id;
    long double timestamp;
};
```

The data reader is instead receiving the message and printing the elapsed time between the current time and the time saved in the timestamp field of the received message. Its callback does basically the following

```
auto received_msg = (RoundTripTimeMsg *)msg;
auto now = std::chrono::duration_cast<std::chrono::nanoseconds>(std::chrono::high_resolution_clock::now().time_since_epoch()).count();
std::cout << "Received message with id: " << received_msg->id() << " and delta time: " << (now - received_msg->timestamp())/1000000.0 << " ms" << std::endl;
```

Both the data writer and the data reader use their default QoS.

I have noticed that if inside the while loop I reduce the workload (using a lower number in the condition) the data reader receives a message earlier. If I increase the workload, the message is received later. What is it happening?

### Fast DDS version/commit

v3.1.0

### Platform/Architecture

Other. Please specify in Additional context section.

### Transport layer

UDPv4

### Additional context

Platform: Ubuntu Jammy Jellyfish 22.04 amd64

**Comment by EugenioCollado** [comment]
Hi @MMarcus95 ,

We have replicated this same context you are describing and have not encountered any latency. Based on your description, the observed latency behavior is likely due to CPU resource contention and not an actual DDS-related issue. The workload simulation inside the DataWriter's loop can monopolize the CPU core it is running on, potentially delaying the other Fast DDS middleware threads if they are running on the same core.

To better understand the cause of the issue, please provide additional details about your system’s configuration, including the number of CPU cores, and observe the CPU core usage during your tests.

Thank you!

**Comment by MMarcus95** [comment]
Hi @EugenioCollado,

thanks for the feedback.

I'm running the test I was talking about inside a Docker Image with Ubuntu 22.04. The CPU is an Intel 13th Gen i7-13700H. In the following there is a screenshot of the _lscpu_ command
![Screenshot from 2024-12-19 14-26-07](https://github.com/user-attachments/assets/5f1dfd61-620c-4e44-9407-97a216773300)

This is the CPU core usage when using the same workload I was mentioning before (so the condition number for the for loop is **10000000**)
![Screenshot from 2024-12-19 15-04-53](https://github.com/user-attachments/assets/78091b4b-9ec6-494f-81f5-87e5c8c8de7c)

In this other example instead, I'm reducing the workload, having as condition number of the for loop **100000**
![Screenshot from 2024-12-19 15-06-28](https://github.com/user-attachments/assets/228c1ed3-1792-4a36-85e0-cfc5f5aa915a)

In both cases, the sender thread is indeed using around 100% of CPU.

However, I see that the message latency is much lower in the second case, as you can see in this plot
![Screenshot from 2024-12-19 15-08-56](https://github.com/user-attachments/assets/57cbcc1e-101e-460d-b5a0-551ab74dd909)


Please let me know if you need any other information.

Thanks!

**Comment by EugenioCollado** [comment]
Hi @MMarcus95 ,

After further investigation, we have replicated the behavior you described. It appears that the issue is not directly related to CPU workload but rather to the publication frequency itself. In your case, the workload in the loop was consuming time, effectively reducing the publication frequency. This can be checked by replacing the workload simulation with sleep calls and seeing the issue persists.

We believe the observed "latency" is a result of how the kernel and its scheduler manage processes. Specifically, the process might be removed from the core's cache and subsequently reloaded, introducing delays.

Our tests show that actions such as isolating a core, forcing the process to run on that core, increasing its priority, and similar mechanisms to prevent the CPU from reloading the process when the publisher is called significantly improve the situation. We would appreciate it if you could test these actions on your side and share your feedback on the results. This would help confirm our conclusions and ensure that these approaches effectively mitigate the issue in your environment.

That said, since this behavior stems from kernel-level process management and not from DDS itself, it is not currently on our roadmap to implement changes addressing this issue in Fast DDS. However, if you'd like, you are welcome to reach out to us directly, and we can explore potential solutions tailored to your specific use case.

Thank you!

**Comment by MMarcus95** [comment]
Hi @EugenioCollado,

thanks for useful information.

I have done similar tests and these are the results under the "workload condition" 10000000:
![results2](https://github.com/user-attachments/assets/90e49d21-d62b-4d2a-a4d8-cc0be4c2cad2)

Let me explain the legend:
- **default: no RT**: running the process as in the previous example, no real-time settings
- **main process-SCHED_FIFO: priority=99, affinity=4**: process scheduled using real-time FIFO policy wiht priority 99 and affinity 4
- **main process-SCHED_DEADLINE: 20,15,20ms**: process scheduled using real-time DEADLINE policy with period=20ms,runtime=15ms,deadline=20ms
- **main process-SCHED_DEADLINE: 200,15,200ms**: process scheduled using real-time DEADLINE policy with period=200ms,runtime=15ms,deadline=200ms

In these test I was also using a PREEMPT-RT patch for the linux kernel.
Setting the real-time FIFO policy seems to decrease the mean "latency". However, when looking at the real-time DEADLINE policy, it seems to be clear that, the higher the publication frequency, the higher the "latency". If I further increase the period of the DEADLINE policy, it seems to keep that "maximum" average "latency", like if there is an upper limit on the maximum "latency".

I also tried to set FIFO scheduler policy to the dds.asyn and dds.udp threads setting respectively `participant_qos.builtin_controllers_sender_thread()` and `udp_transport->default_reception_threads(eprosima::fastdds::rtps::ThreadSettings{scheduling_policy,priority,affinity,stack_size});`, but the results  did not change much.

**Comment by EugenioCollado** [comment]
Hi @MMarcus95 ,

Thank you for providing such detailed feedback and for conducting thorough tests. We appreciate your effort in exploring different configurations and sharing the results. It's great to see that  the latency has been reduced with those actions.

Given the relevance of this subject and the potential interest it may hold for other users, I will move this issue to the discussions section.

Thank you again for your feedback!
